ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -345,4 +386,148 @@ pub mod tests {
+            for i in 0..self[0].len() {
+                let elem = [self[0][i].clone(), self[1][i].clone(), self[2][i].clone()];
+                res.push(elem.reconstruct());
+            }
+            res
+        }
+    }
+
+    const TEST_USER_IDS: [usize; 8] = [1, 2, 3, 5, 8, 13, 21, 34];
+
+    // test for sorting multiple ranges in a longer list
+    #[test]
+    fn test_multiple_ranges() {
+        run(|| async move {
+            let world = TestWorld::default();
+            let mut rng = thread_rng();
+
+            // test cases for both, ascending and descending
+            let bools = vec![false, true];
+
+            for desc in bools {
+                // generate vector of structs corresponding to 8 users.
+                // Each user will have a different number of records
+                // Each struct will have a random timestamps
+                let mut records: Vec<SillyStruct> = Vec::with_capacity(TEST_USER_IDS.iter().sum());
+                for user_id in TEST_USER_IDS {
+                    for _ in 0..user_id {
+                        records.push(SillyStruct {
+                            timestamp: rng.gen::<BA20>(),
+                            user_id,
+                        });
+                    }
+                }
+
+                // convert expected into more readable format
+                let mut expected: Vec<(usize, u128)> = records
+                    .clone()
+                    .into_iter()
+                    .map(|x| (x.user_id, x.timestamp.as_u128()))
+                    .collect();
+                // sort expected
+                expected.sort_unstable_by(|a, b| match a.0.cmp(&b.0) {
+                    Ordering::Less => Ordering::Less,
+                    Ordering::Greater => Ordering::Greater,
+                    Ordering::Equal => {
+                        if desc {
+                            b.1.cmp(&a.1)
+                        } else {
+                            a.1.cmp(&b.1)
+                        }
+                    }
+                });
+
+                let (_, ranges_to_sort) = TEST_USER_IDS.iter().fold(
+                    (0, Vec::with_capacity(TEST_USER_IDS.len())),
+                    |acc, x| {
+                        let (start, mut ranges) = acc;
+                        let end = start + x;
+                        ranges.push(start..end);
+                        (end, ranges)
+                    },
+                );
+
+                // compute mpc sort
+                let result: Vec<_> = world
+                    .semi_honest(records.into_iter(), |ctx, mut records| {
+                        let ranges_clone = ranges_to_sort.clone();
+                        async move {
+                            quicksort_ranges_by_key_insecure(
+                                ctx,
+                                &mut records[..],
+                                desc,
+                                |x| &x.timestamp,
```

Comment:
You could make this a method on SillyStruct and use SillyStruct::timestamp, which is a pattern I would encourage real code to use.